### PR TITLE
Allow images with singleton dimensions for regionprops

### LINKED ISSUE #2105: regionprops dimensions of one and single element regions
Hi,

skimage.measure.regionprops does not handle images with an axis having a of size 1 correctly. E.g. an image of shape (1,3), which has a single row, is not detected as an 2D-image and an error is thrown, which says, that only 2D and 3D image are allowed.
Same for 3D images of e.g. shape (1,1,3).

Further, properties of regions with only one pixel are not correct. At least the property 'major_axis_length' results in 0.0 for a single pixel region (guess it should be the diagonal of the pixel: sqrt(2)).

### PR DESCRIPTION
**Pull request by jni**
Fixes #2105.

I think the ultimate goal should be to not use squeeze, because it will result in unpredictable behaviour in the future. But the deprecation path is tricky, so help is appreciated here.

### CODE REVIEW

**Comment by JDWarner** [comment]
What was the reason for `squeeze` to begin with? To remove a possible singleton channel representing gray values?

Also, the `regionprops` docs don't say anything about 3D inputs.

**Comment by jni** [comment]
@JDWarner the choice of squeeze remains a mystery to me, but your hypothesis rings true. I would personally be happy to remove it altogether, but I imagine that will result in user complaints come 0.13.

**Comment by JDWarner** [comment]
IMO this was at best undocumented behavior, at worst a bug. I lean toward bug. Removing dimensions at will isn't something that should be taken lightly.

As a bridge for anyone who came to depend on this "feature", we could 

* Raise an informative error if the input is 4D *and* includes a singleton dimension. Tell the user that we don't assume this anymore; a volumes with a "dummy channel" are no longer supported. Yes, it breaks, but they get told how to fix it.
* Warn any users with 3D inputs that include a singleton dimension that their input was assumed to be a 3D volume, not a 2D image with a "dummy channel". If they wanted the 2D behavior, *they* should apply `np.squeeze` or slice that channel out of a multichannel array.

The burden should be on the user to provide the right input.

File: skimage/measure/_regionprops.py [hunk 1]
```diff
@@ -517,10 +517,13 @@ def regionprops(label_image, intensity_image=None, cache=True):
...
     """
 
-    label_image = np.squeeze(label_image)
+    squeezed_label_image = np.squeeze(label_image)
 
-    if label_image.ndim not in (2, 3):
-        raise TypeError('Only 2-D and 3-D images supported.')
+    if squeezed_label_image.ndim not in (2, 3):
+        if label_image.ndim not in (2, 3):
+            raise TypeError('Only 2-D and 3-D images supported.')
```
Comment (soupault) [inline skimage/measure/_regionprops.py]: I guess the signature should be changed accordingly. At the moment, it is: `label_image : (N, M) ndarray`.

**Comment by soupault** [comment]
> Also, the regionprops docs don't say anything about 3D inputs.

Exactly. Neither it states that multichannel input is valid.

I suggest to remove the `np.squeeze` call completely and update all the methods to support the arrays with singleton dimensions.

> Warn any users with 3D inputs that include a singleton dimension that their input was assumed to be a 3D volume, not a 2D image with a "dummy channel". If they wanted the 2D behavior, they should apply np.squeeze or slice that channel out of a multichannel array. Remove this warning after 1-2 major point releases.

From my point of view, we should put all this thing into the docstring: specify correct parameters' signatures and mention that (M,  1) and (M, ) are not considered as equal.

**Comment by emmanuelle** [comment]
+1 for raising error / warning as suggested by @JDWarner . @jni how does it sound?

**Comment by jni** [comment]
@emmanuelle I'm happy to do this! I'd like to know whether @stefanv wants to be the voice of caution here. ;)

**Comment by ahojnnes** [comment]
@jni @stefanv Ping.

**Comment by rfezzani** [comment]
I see [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_regionprops.py#L652) that `np.squeeze` is no more used. Can we close @jni?

**Comment by jni** [comment]
Yep, looks like someone scooped me! In open source that's the best thing. =D

**Pull request closed without merging**